### ISSUE TITLE
PG is not working past picking the 1st card

### ISSUE THREAD

**Issue by Vitae666** [MatthieuBonne/CQAutomater#16]
[ActionLog.txt](https://github.com/MatthieuBonne/CQAutomater/files/3499570/ActionLog.txt)
[ErrorLog.txt](https://github.com/MatthieuBonne/CQAutomater/files/3499571/ErrorLog.txt)
[ExceptionLog.txt](https://github.com/MatthieuBonne/CQAutomater/files/3499572/ExceptionLog.txt)
Added the ExceptionLog because it crashed. Not sure if it was because it couldn't open the ErrorLog which was getting hammered with the PG Error

**Comment by Vitae666** [comment]
Also, might need to someone have it know when the game is over. Apparently it's still trying to send a click on the matching pair that I had left
8/13/2019 10:27:53 PM
	Error Cloud Script Error: Send PGEvent {"ok":false,"err":"Thank you for playing"}
8/13/2019 10:28:19 PM
8/13/2019 10:28:19 PM
8/13/2019 10:28:21 PM
	Error Cloud Script Error: Send PGEvent {"ok":false,"err":"Thank you for playing"}
8/13/2019 10:28:25 PM
	Error Cloud Script Error: Send PGEvent {"ok":false,"err":"Thank you for playing"}
8/13/2019 10:28:49 PM
8/13/2019 10:28:49 PM
	Error Cloud Script Error: Send PGEvent {"ok":false,"err":"Thank you for playing"}
8/13/2019 10:28:49 PM
8/13/2019 10:28:53 PM
	Error Cloud Script Error: Send PGEvent {"ok":false,"err":"Thank you for playing"}
8/13/2019 10:29:19 PM
8/13/2019 10:29:21 PM
	Error Cloud Script Error: Send PGEvent {"ok":false,"err":"Thank you for playing"}
8/13/2019 10:29:22 PM
8/13/2019 10:29:25 PM
	Error Cloud Script Error: Send PGEvent {"ok":false,"err":"Thank you for playing"}

8/13/2019 10:27:51 PM
	Debug in PGCards, pick 10 and 12
8/13/2019 10:28:18 PM
	 Picking PG cell 10
8/13/2019 10:28:19 PM
	 Picking PG cell 12
8/13/2019 10:28:19 PM
	Debug in PGCards, pick 10 and 12
8/13/2019 10:28:48 PM
	 Picking PG cell 10
8/13/2019 10:28:49 PM
	 Picking PG cell 12
8/13/2019 10:28:49 PM
	Debug in PGCards, pick 10 and 12
8/13/2019 10:29:18 PM
	 Picking PG cell 10
8/13/2019 10:29:19 PM
	 Picking PG cell 12
8/13/2019 10:29:22 PM
	Debug in PGCards, pick 10 and 12
8/13/2019 10:29:48 PM
	 Picking PG cell 10
8/13/2019 10:29:49 PM
	 Picking PG cell 12

Same output in the ExceptionLog as well because I crashed.

Maybe add a "Disable Auto-Events" checkbox? If it's going to crash this often, won't help much right now :(
It was crashing every 5 or so mins, but now it's on 10mins and nothing. Honestly, don't know why or if it even was 100% the reason, it would crash because it couldn't write to Errorlog.

**Comment by MatthieuBonne** [comment]
Ouch :\ Do the pg event manually today please. I try to fix that, dunno why my test worked last week.

**Comment by MatthieuBonne** [comment]
Checkbox added. Hope we'll end this work next PGE day. :)

**Comment by Vitae666** [comment]
Thanks. Yeah, figured the checkbox might be a good thing JUST in case things go to hell in a handbasket. So it could be switched off. 
I think CQ Devs eventually might notice 6 attempts a minute at picking a cell when the game is over LOL!
Not that CQA is a hidden thing. They know all about it. But still. Discretion is a better part of saving your ass from the banhammer ;-)

**Comment by Vitae666** [comment]
PSSST.....Checkbox doesn't save between sessions. Closed it down, reopened it, and had to recheck it :)
Yeah...I'm that lazy ROFL!

**Comment by MatthieuBonne** [comment]
Nice catch, fixed !

**Comment by Vitae666** [comment]
I R asmrt

**Comment by Vitae666** [comment]
Just noticed that it was Match Pair Event. Don't think CQA knew it either.
I did some manual picks. Closed CQA. Restarted it. And it started picking the 1 matching pair I had, then it froze up again :(

8/22/2019 10:24:09 AM
	Debug in PGCards, PGCards value 6
8/22/2019 10:24:09 AM
	 Picking PG cell 0
8/22/2019 10:24:10 AM
	 Picking PG cell 3
8/22/2019 10:24:11 AM
	Debug in PGCards, pick 0 and 3
8/22/2019 10:24:39 AM
	Debug in PGCards, PGCards value 4
8/22/2019 10:24:39 AM
	 Picking PG cell 1
8/22/2019 10:24:40 AM
	 Picking PG cell 1
8/22/2019 10:25:09 AM
	Debug in PGCards, PGCards value 3
8/22/2019 10:25:09 AM
	 Picking PG cell 1
8/22/2019 10:25:10 AM
	 Picking PG cell 1
8/22/2019 10:25:39 AM
	Debug in PGCards, PGCards value 3
8/22/2019 10:25:39 AM
	 Picking PG cell 1
8/22/2019 10:25:40 AM
	 Picking PG cell 1
8/22/2019 10:26:09 AM
	Debug in PGCards, PGCards value 3
8/22/2019 10:26:09 AM
	 Picking PG cell 1
8/22/2019 10:26:09 AM
	 Picking PG cell 1
8/22/2019 10:26:39 AM
	Debug in PGCards, PGCards value 3
8/22/2019 10:26:39 AM
	 Picking PG cell 1
8/22/2019 10:26:39 AM
	 Picking PG cell 1
8/22/2019 10:27:09 AM
	Debug in PGCards, PGCards value 3
8/22/2019 10:27:09 AM
	 Picking PG cell 1
8/22/2019 10:27:09 AM
	 Picking PG cell 1

And I'm thinking that when it's picking a card that already has a match, the game itself is screwing up and when you do another manual pick, it thinks you are trying to match those 2 cards together.
I didn't get all my picks before the game ended.
With 8 picks that's a 2 whole rows. I didn't get that. So, there's a bug in game as well.
What I mean here, is that CQA was picking #3 and I was clicking cards and it was taking away attempts.
Not sure if it's an in game bug, or if it's a backend thing where CQA is selecting the card where clicking matched cards in-game are not possible.
Basically, it was my fault for not shutting down CQA while I was doing the manual clicks. BUT I'm not sure if there's a bug in game as well.
![image](https://user-images.githubusercontent.com/9310914/63523346-fb96a280-c4c7-11e9-8af1-588149874c53.png)

The good news is that CQA knew when there were no more Attempts to be made and disabled itself :)

**Comment by MatthieuBonne** [comment]
There's still some bugs to fix, but yes, doing manual clicks while CQA is running will mess up things. I try to improve the code til next PG event.

**Comment by Vitae666** [comment]
I feel like now it's not even knowing that it's PG day. It's been sitting at "Not active today" the whole day. I clicked #1 and #2 and suddenly it woke up. 
BUT, with the same errors. Except a hell of a lot worse....
8/30/2019 12:26:22 AM
	Debug in PGCards, PGCards attempts remaining : 3
8/30/2019 12:26:22 AM
	 Picking PG cell 0
8/30/2019 12:26:27 AM
	 Picking PG cell 1
8/30/2019 12:26:48 AM
	 Sending coupon 33PNwUxbzG
8/30/2019 12:26:54 AM
	 Coupon gave 1 Key
8/30/2019 12:26:55 AM
	Debug in PGCards, PGCards attempts remaining : 2
8/30/2019 12:26:55 AM
	 Picking PG cell 0
8/30/2019 12:26:57 AM
	 Picking PG cell 1
8/30/2019 12:27:26 AM
	Debug in PGCards, PGCards attempts remaining : 1
8/30/2019 12:27:26 AM
	 Picking PG cell 0
8/30/2019 12:27:45 AM
	 Picking PG cell 1

![image](https://user-images.githubusercontent.com/9310914/63993255-a1946f00-cabd-11e9-857f-33ca5b42670e.png)

**Comment by MatthieuBonne** [comment]
Jeez, I didn't think this event would be so much harder than the other ones. Sorry, trying to fix that.

**Comment by Vitae666** [comment]
I know 😀
You've added some really great new things that I didn't think were possible.
I'm not gonna get pissy cause 1 of 10 don't work right 

BTW, Fuck yeah on the coupons! I swear I was going to ask you about it last week, but I thought "oh hellz no. From one website to another? I'm asking to much!!"

**Comment by MatthieuBonne** [comment]
Thank you. ^^
I just sent an update, I think I found my mistake, and it worked on my account. Wait & see...

**Comment by Vitae666** [comment]
BTW, DOES the Twitter thing do the key collect or does it just show you what the new coupon is? Just wondering cause it didn't collect on it today. Not even 100% sure it has in the past either since being put in.

**Comment by MatthieuBonne** [comment]
I had disabled it to debug issue #19, sorry, I reactivated it.

**Comment by Vitae666** [comment]
Ok, so I have 0 attempts in PG but I have nothing in the logs as to having attempted to do it.
The heck?

EDIT: I mean 0 attempts and no cards flipped at all.

**Comment by Vitae666** [comment]
Ok, this is rubbish on CQ's side, not CQA. It showed 0 Attempts. I decided, "What would happen if I clicked a card?" soooo...I did. And immediately got the 8 Attempts and CQA picked it up from there.
Any way to have CQA notice that when PG (WHY PG when it should be MP for Matching Pairs, BTW??) know to force a pick when the Event 1st becomes available?
I don't know if it's a possible error that could happen at the end when it hits 0 again that it will try to pick over and over?
Able to mitigate that by keeping a "PreviousAttemptCount" variable? So that it can read
CurrentAttemptCount == 0 && PreviousAttemptCount = 1
Engage PickScript
Else EndPGEvent

Obviously, I have no clue what you have things called in there, but that's pretty much what I mean :)

Other than that, love how it's working now!

**Comment by MatthieuBonne** [comment]
Hmmm, mine started well and played fine. There's a little issue where CQA tries to play PG on non-PG day, gotta fix that !

**Comment by Vitae666** [comment]
Haven't seen it try to play on a non-pg day.
I check the logs semi regularly to know what the heck is going on behind the scenes :)

EDIT: Let me amend that a bit. While I don't see it trying to PLAY PG, I do see it saying that the "event completed today" even days after it's been over.
But I've never actually noticed it try to pick cards.

**Comment by Vitae666** [comment]
PG didn't start yet for me.

> 9/15/2019 12:37:59 AM
> 	 Sending coupon hIUTXJBgdT
> 9/15/2019 12:38:02 AM
> 	 Coupon gave 1 Key
> 9/15/2019 12:47:49 AM
> 	FAILED Bid on hero Reindeer for: 125UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 12:47:50 AM
> 	FAILED Bid on hero Reindeer for: 125UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 12:48:43 AM
> 	FAILED Bid on hero Reindeer for: 138UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 12:48:48 AM
> 	FAILED Bid on hero Reindeer for: 138UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 12:49:52 AM
> 	FAILED Bid on hero Reindeer for: 152UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 12:49:54 AM
> 	FAILED Bid on hero Reindeer for: 152UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 12:50:46 AM
> 	FAILED Bid on hero Reindeer for: 168UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 12:50:52 AM
> 	FAILED Bid on hero Reindeer for: 168UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 12:51:44 AM
> 	FAILED Bid on hero Reindeer for: 185UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 12:51:48 AM
> 	FAILED Bid on hero Reindeer for: 185UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 12:52:47 AM
> 	FAILED Bid on hero Reindeer for: 204UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 12:52:50 AM
> 	FAILED Bid on hero Reindeer for: 204UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 12:56:22 AM
> 	 Sending coupon hIUTXJBgdT
> 9/15/2019 10:26:26 AM
> 	 Sending coupon hIUTXJBgdT
> 9/15/2019 11:09:19 AM
> 	SUPER LORD OF CHAOS Heroes Allowed fought with: athert:99 thert:99 aurora:19 raiderrose:99 W31 ageum:99
> 9/15/2019 11:09:22 AM
> 	SUPER LORD OF CHAOS Heroes Allowed fought with: athert:99 thert:99 aurora:19 raiderrose:99 W31 ageum:99
> 9/15/2019 11:10:16 AM
> 	SUPER LORD OF CHAOS Heroes Allowed fought with: athert:99 thert:99 aurora:19 raiderrose:99 W31 ageum:99
> 9/15/2019 11:13:14 AM
> 	Leveled up hero: Liliwith:AS
> 9/15/2019 11:32:21 AM
> 	Leveled up hero: Alordkirkwith:CC
> 9/15/2019 11:35:37 AM
> 	Leveled up hero: Alordkirkwith:CC
> 9/15/2019 12:38:58 PM
> 	 Sending coupon hIUTXJBgdT
> 9/15/2019 12:43:57 PM
> 	Bid on hero Hidoka for: 267UM.
> 9/15/2019 12:44:01 PM
> 	FAILED Bid on hero Hidoka for: 267UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 1:13:56 PM
> 	FAILED Bid on hero Reindeer for: 301UM.
> {"ok":false,"err":"Not enough UM"}
> 9/15/2019 1:14:00 PM
> 	FAILED Bid on hero Reindeer for: 301UM.
> {"ok":false,"err":"Not enough UM"}
> Got Crei
> 
> 9/15/2019 8:38:57 PM
> 	 Sending coupon DKwcczmUNg
> 9/15/2019 8:38:58 PM
> 	 Coupon gave 1 Key
> 

It's now 11:52 PM
:-/

**Comment by Vitae666** [comment]
Take a look at what I mean. The opening screen is supposed to be my clock. Stupid screen recorder did something weird when I started the recording. Had to switch from my Screencast-O-Matic to BSR Screen Recorder because SoM has stopped working for some reason.

[https://youtu.be/SGaNvTS8sU0](https://youtu.be/SGaNvTS8sU0)

For the record, before I did the recording, I'd already enabled and disabled Events as well as restarted CQA to see if either of them would get it to start. Obviously, none of that did :)

Also, here's the dump from ActionLog just cause I have no idea what the Debugs are about.

9/16/2019 12:32:16 AM	 Picking PG cell 1
9/16/2019 12:32:17 AM	 Picking PG cell 2
9/16/2019 12:32:48 AM	 Picking PG cell 3
9/16/2019 12:32:49 AM	 Picking PG cell 4
9/16/2019 12:33:15 AM	 Picking PG cell 5
9/16/2019 12:33:16 AM	 Picking PG cell 4
9/16/2019 12:33:45 AM	 Picking PG cell 6
9/16/2019 12:33:46 AM	 Picking PG cell 1
9/16/2019 12:34:16 AM	 Picking PG cell 7
9/16/2019 12:34:17 AM	 Picking PG cell 8
9/16/2019 12:34:45 AM	 Picking PG cell 9
9/16/2019 12:34:46 AM	 Picking PG cell 3
9/16/2019 12:35:16 AM	 Picking PG cell 10
9/16/2019 12:35:17 AM	 Picking PG cell 11
9/16/2019 12:35:45 AM	 Picking PG cell 12
9/16/2019 12:35:47 AM	 Picking PG cell 13
9/16/2019 12:36:16 AM	 Picking PG cell 14
9/16/2019 12:36:17 AM	 Picking PG cell 2
9/16/2019 12:36:45 AM	 Picking PG cell 15
9/16/2019 12:36:46 AM	 Picking PG cell 16
9/16/2019 12:37:16 AM	 Picking PG cell 16
9/16/2019 12:37:17 AM	 Picking PG cell 17
9/16/2019 12:37:46 AM	 Picking PG cell 15
9/16/2019 12:37:48 AM	 Picking PG cell 17
9/16/2019 12:37:49 AM	Debug in PGCards, pick 15 and 17
9/16/2019 12:38:17 AM	 Picking PG cell 18
9/16/2019 12:38:20 AM	 Picking PG cell 19
9/16/2019 12:38:46 AM	 Picking PG cell 16
9/16/2019 12:38:47 AM	 Picking PG cell 19
9/16/2019 12:38:49 AM	Debug in PGCards, pick 16 and 19
9/16/2019 12:39:16 AM	 Picking PG cell 20
9/16/2019 12:39:17 AM	 Picking PG cell 12
9/16/2019 12:39:46 AM	 Picking PG cell 21
9/16/2019 12:39:47 AM	 Picking PG cell 0
9/16/2019 12:40:16 AM	 Picking PG cell 22
9/16/2019 12:40:20 AM	 Picking PG cell 18
9/16/2019 12:40:46 AM	 Picking PG cell 23
9/16/2019 12:40:51 AM	 Picking PG cell 8
9/16/2019 12:41:16 AM	 Picking PG cell 24
9/16/2019 12:41:17 AM	 Picking PG cell 25

**Comment by MatthieuBonne** [comment]
Weird it didn't start alone. Trying a fix in next commit.

**Comment by Vitae666** [comment]
Yeah,dunno why. I know you said yours started last week just fine. No idea why it's giving me grief.

**Comment by MatthieuBonne** [comment]
Please tell me tomorrow how it goes ! :)

**Comment by Vitae666** [comment]
You know I will 🤣

**Comment by Vitae666** [comment]
_**EDIT**: All the below is on the latest version. Forgot to mention that. And no, at no point did it cross my mind to check to see what would happen on the previous version. I manually completed MC._

WEEEEEELLLLLLLL.......
![image](https://user-images.githubusercontent.com/9310914/66174021-3dffe300-e621-11e9-97ab-94a7e61b4d87.png)

And about 30 seconds after manually I pick the 1st card:
![image](https://user-images.githubusercontent.com/9310914/66174119-756e8f80-e621-11e9-9783-293013b4249e.png)

And 30 minutes later:
![image](https://user-images.githubusercontent.com/9310914/66175128-70f8a580-e626-11e9-8022-43260ed5269e.png)

```
10/3/2019 9:04:10 PM	 Picking PG cell 1
10/3/2019 9:04:42 PM	 Picking PG cell 1
10/3/2019 9:05:13 PM	 Picking PG cell 1
10/3/2019 9:05:45 PM	 Picking PG cell 1
10/3/2019 9:06:17 PM	 Picking PG cell 1
...
10/3/2019 9:34:36 PM	 Picking PG cell 1
10/3/2019 9:35:08 PM	 Picking PG cell 1
10/3/2019 9:35:40 PM	 Picking PG cell 1
10/3/2019 9:36:12 PM	 Picking PG cell 1
```

Note that there's nothing even in the ActionLog that says it clicked cell 2. And I sure didn't do it :)
Also note that cell 2 is highlighted and I can't undo it. I've tried clicking it. Closing the game and browser down. Nothing. It restarts with it selected.
But as soon as I manually pick the next cell it's an automatic Attempt use.
![image](https://user-images.githubusercontent.com/9310914/66175205-b3ba7d80-e626-11e9-8216-3aeebff1bf0f.png)

I disabled Events. Closed CQA. Waited a bit. Restarted it. Enabled Events. Waited.
And again, cell 2 is highlighted and no mentioned in the log about it.

**Comment by Vitae666** [comment]
Question: 
Which method do you think is better? 
Clicking all the cells until you have 1 attempt left and THEN making all possible matches with whats turned over, then clicking 1 cell on the last attempt and if it matches one, doing that, obviously, or if not, just whatever last cell?
Or actually going for the matches as they pop up?
I'm not sure if there's any better reason for doing 1 over the other. Like I don't think you'd get any less or more matches because the matches would be displayed regardless.
Do you think having MC (Still not calling it PG) would benefit and have less errors by clicking along until there's 1 match left and then doing "the needful" from there?
Not sure if this oddness is happening because it's looking for matches right away or what. Again, as always, what the hell do I know? :)

**Comment by MatthieuBonne** [comment]
Jeez... Tried another way to fix it.
It's better to use matches asap I guess. And there are only 2 ways to win (i.e. spare attempts) : 
- do 2 random picks and discover they're twins (woohoo, lucky)
- have your first pick (from a try of 2) be random AND match an existing card.
If only the second pick is a match, you'll win pg next round (if any left), but that's all. No way to get a nice score without those two lucky events.

**Comment by Vitae666** [comment]
Yeah, when I do it manually, I just click all the way until the last attempt and then take care of the matches. Then just flip 1 card and see if it's a match.
Just wasn't sure if that was the best way. But I guess, like I said, it won't hurt your chances. It shouldn't lessen the chance. I think.

**Comment by Vitae666** [comment]
W00T! Went off without a hitch this time!
